### PR TITLE
rustdoc: Remove `Crate.externs` and compute on-demand instead

### PR DESCRIPTION
**Pull request by camelid**
r? @GuillaumeGomez 
cc @jyn514

### CODE REVIEW

File: src/librustdoc/clean/utils.rs [hunk 1]
```diff
@@ -24,17 +24,8 @@ use std::mem;
...
 crate fn krate(cx: &mut DocContext<'_>) -> Crate {
-    use crate::visit_lib::LibEmbargoVisitor;
-
     let module = crate::visit_ast::RustdocVisitor::new(cx).visit();
 
-    let mut externs = Vec::new();
-    for &cnum in cx.tcx.crates(()) {
-        externs.push(ExternalCrate { crate_num: cnum });
-        // Analyze doc-reachability for extern items
-        LibEmbargoVisitor::new(cx).visit_lib(cnum);
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:35]: Are you sure changing the time at which this pass is run has no effect? I'd believe anything of rustdoc 😅

**Review by jyn514 — COMMENTED** [review]
Lol you'd already fixed my one comment by the time I finished the review, nice

**Comment by jyn514** [comment]
Anyway, this looks good whether or not it's a perf improvement, r=me

File: src/librustdoc/clean/utils.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/clean/utils.rs:35]: Yep, it definitely has an effect! As you said, it's already fixed.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
It seems like it doesn't impact performance or memory usage. I'm a bit surprised but doesn't matter. Thanks for the simplification!

@bors: r=jyn514,GuillaumeGomez

**Comment by mathstuf** [comment]
It looks like this is probably the cause of a regression. See #91134.

**Comment by camelid** [comment]
Update from that issue: the regression is likely from somewhere else.

**Pull request merged**